sql/core/src/main/scala/org/apache/spark/sql/execution/command/views.scala

```diff
@@ -88,23 +96,23 @@ case class CreateViewCommand(
     qe.assertAnalyzed()
     val analyzedPlan = qe.analyzed
 
-    if (tableDesc.schema != Nil && tableDesc.schema.length != analyzedPlan.output.length) {
+    if (userSpecifiedColumns.nonEmpty &&
+        userSpecifiedColumns.length != analyzedPlan.output.length) {
       throw new AnalysisException(s"The number of columns produced by the SELECT clause " +
         s"(num: `${analyzedPlan.output.length}`) does not match the number of column names " +
-        s"specified by CREATE VIEW (num: `${tableDesc.schema.length}`).")
+        s"specified by CREATE VIEW (num: `${userSpecifiedColumns.length}`).")
     }
     val sessionState = sparkSession.sessionState
 
     if (isTemporary) {
-      createTemporaryView(tableDesc.identifier, sparkSession, analyzedPlan)
+      createTemporaryView(sparkSession, analyzedPlan)
     } else {
       // Adds default database for permanent table if it doesn't exist, so that tableExists()
       // only check permanent tables.
-      val database = tableDesc.identifier.database.getOrElse(
-        sessionState.catalog.getCurrentDatabase)
-      val tableIdentifier = tableDesc.identifier.copy(database = Option(database))
+      val database = name.database.getOrElse(sessionState.catalog.getCurrentDatabase)
+      val qualifiedName = name.copy(database = Option(database))
 
-      if (sessionState.catalog.tableExists(tableIdentifier)) {
+      if (sessionState.catalog.tableExists(qualifiedName)) {
```

Comment:
The existing code sounds like a bug. It does not check whether the existing table is a table or a view. Will fix it in a separate PR.